### ISSUE TITLE
RFC: spatial: distance computation kernels

### ISSUE THREAD

**Issue by fancidev** [scipy/scipy#21645]
### Is your feature request related to a problem? Please describe.

TBC

### Describe the solution you'd like.

TBC

### Describe alternatives you've considered.

TBC

### Additional context (e.g. screenshots, GIFs)

TBC

**Comment by czgdp1807** [comment]
This is interesting actually. I will go through your comment and infact would like to take this up. Let's meet on call to discuss this (if its okay to meet). 

cc: @rgommers

**Comment by lucascolley** [comment]
There is overlap between the kernel dispatching discussed here and past work on Uarray: https://github.com/scipy/scipy/issues/14353. Uarray work seems to have paused so it may be wise to hold fire before we start reinventing the wheel. See also https://github.com/scientific-python/spatch/issues/1.

**Comment by fancidev** [comment]
> There is overlap between the kernel dispatching discussed here and past work on Uarray: #14353. Uarray work seems to have paused so it may be wise to hold fire before we start reinventing the wheel. See also [scientific-python/spatch#1](https://github.com/scientific-python/spatch/issues/1).

Thanks for the links! They are interesting to read.

I think this proposal is quite different from the referenced ones in that it is (1) less ambitious, and (2) more concrete.

For (1), the proposal deals exclusively with dense arrays stored in main memory and eagerly computed on the CPU. This is exactly what the current distance computation code in SciPy handles. The limited scope makes the proposal an incremental enhancement rather than a complete rewrite, making it easier to materialize. For example, GPU computation or alternative array backends are completely out of the scope of this proposal.

For (2), the proposal addresses a specific issue — performance of distance computation (for dense arrays on the CPU). To optimize performance, specialized routines are needed for each combination of (metric, CPU, data type); the number of combinations can easily reach hundreds and become unmanageable by the limited developer resource of SciPy. In fact, SciPy’s distance functions currently only support `f64` input, which might have rendered themselves irrelevant for applications that operate on large amount of low precision data. This proposal makes an attempt to make at least the API relevant.

It might be helpful to compare the proposed kernel interface to BLAS — both define a concrete set of operations, for which efficient implementations are provided by some backend and consumed by some frontend.

**Comment by lorentzenchr** [comment]
@jjerphan @Micky774 your inout might be valuable.

I would recommend to invest some effort into an analysis of matches and difference between `scipy.spatial.distance` and `sklearn.metrics.pairwise`. No need to have 2 largely overlapping implementations of the same functionality.

Then about performance: The 3 main drivers are cache efficiency, use of simd, use of multithreading.

**Comment by jjerphan** [comment]
Thank you for the heads-up, @lorentzenchr.

This RFC is indeed valuable since there's some overlap in the ecosystem and unaddressed issues for distance metrics.

I am genuinely interested in participating actively to the discussions, but I lack time for now (I have other commitments which I want to honor first). I can provide a few pointers for issues or PRs on scikit-learn and participate passively when I get some time.

**Comment by jjerphan** [comment]
A list of a few elements which I might update from time to time:
 - Some of the most notable contribution authored for scikit-learn are Meekail's:
    - https://github.com/scikit-learn/scikit-learn/pull/27145 
    - https://github.com/scikit-learn/scikit-learn/pull/26983
 - conda-forge now supports [microarchitecture-optimized builds](https://conda-forge.org/docs/maintainer/knowledge_base/#microarch)

**Comment by fancidev** [comment]
Thanks for the comments!

> No need to have 2 largely overlapping implementations of the same functionality so close in the ecosystem.

From my limited knowledge about both libraries, it appears to me that scipy’s distance functions focus more on 3D vectors, while sklearn’s functions focus more on large vectors with hundreds to thousands of elements.

> Then about performance: The 3 main drivers are cache efficiency, use of simd, use of multithreading.

While this RFC is not about how to improve speed itself (but about how to make it possible to plug in  performant implementations), I could share some comments about the current performance characteristics of scipy’s implementation.

Cache efficiency seems ok as data are mostly read and written in serial (no jumping around). For long vectors (hundreds to thousands elements) the manual unrolling that SciPy implements may read from up to 8 pages at the same time. Depending on the hardware and system load this may improve or degrade performance. Light experiments on my own computer appears to suggest it still improves performance.

SIMD is not explicitly coded but relies on compiler auto-vectorization. I’m not sure SciPy has the resource to hand-tune SIMD intrinsics. [Light experiments] suggest that there’s still room to adjust the compiler flags to improve performance significantly without sacrificing accuracy. Separately, dynamic dispatching might be more distribution friendly, and e.g. SinSIMD already does that.

Multi-threading can be implemented at different levels, and it may be tricky to coordinate. An obvious candidate is OpenMP, but a quick search on GitHub Issues reveals quite a few practical obstacles. Therefore SciPy’s distance functions are single threaded (but thread safe) and I’m not aware of any plan to make them threaded.

**Comment by rgommers** [comment]
Thanks @fancidev! Overall the motivation and ideas for a plugin infrastructure look good to me. It would be great to be able to have a good interface for integration of a package like SimSIMD.

> one or more of the following C interfaces

Regarding the signatures: the most important choice that jumps out to me is that you're choosing to include strides. It may be easier and better for performance for another library to only support contiguous arrays. JAX and Pythran are two examples of packages that made this choice. So it's worth thinking about whether you'd have a contiguous-only interface either in addition to or instead of the strided one. 

> To achieve this, I propose to split the core computation logic of a metric into “distance computation kernels” that implement one or more of the following C interfaces:
> [...]
> To make a kernel available for use by SciPy, the backend would “register” the (metric, in_dtype, out_dtype) combinations it supports by calling the following:

Did you prototype this mechanism? Some important details of how this could work are missing, for example you can't just pass C function pointers to a Python-level registration mechanism. I suspect it should use `scipy.LowLevelCallable` or something similar to it. The closest analog to it we have is http://scipy.github.io/devdocs/tutorial/ndimage.html#extending-scipy-ndimage-in-c.

**Comment by fancidev** [comment]
Thanks for the comments @rgommers. They're very good points!

> So it's worth thinking about whether you'd have a contiguous-only interface either in addition to or instead of the strided one.

My initial objective was to fully support the current SciPy behavior. The current SciPy distance implementation supports arbitrary strides between rows *and* between columns. This ensures zero copy as long as the elements are aligned, though the performance can be suboptimal if the strides are not one.

I made a compromise between generality and performance by including only "row strides" in the signature -- the elements within each vector are still contiguous. I find that the BLAS routines ?GEMM and ?GEMV both accept a stride, so I assume including such a stride has no impact on performance. In addition, SimSIMD uses a double loop to implement `cdist`, and supporting row strides is "free".

But this indeed leaves the interface in an awkward position where it has redundant arguments for the best case (fully contiguous input), and yet it does not support zero-copy in the worst case (where input data has non-zero column stride). 

So I'm going to amend the interface to include a "slim signature" that assumes fully contiguous inputs and a "full signature" that supports arbitrarily strided inputs. Performance centric implementations only need to support the "slim signature", and I expect that should be good enough for most of the real-world performance sensitive tasks.

> Did you prototype this mechanism? Some important details of how this could work are missing, for example you can't just pass C function pointers to a Python-level registration mechanism. I suspect it should use `scipy.LowLevelCallable` or something similar to it. The closest analog to it we have is http://scipy.github.io/devdocs/tutorial/ndimage.html#extending-scipy-ndimage-in-c.

Indeed that "registration" code segment is a sketch. I wasn't aware of `scipy.LowLevelCallable`, let me look into it.

What I did in a prototype was: (1) build SimSIMD into a pure-C DLL (without Python binding), and (2) use `ctypes` in Python to load the DLL, (3) load the exported C function symbol as a `ctypes` function, and (4) use `numpy.ctypes.` to call the function with NumPy array input. This works, but it is clearly quite far from a production grade solution.

**Comment by ashvardanian** [comment]
@fancidev, for context, SimSIMD exposes the function pointers into the Python API as well, and it may be compatible with `scipy.LowLevelCallable`, mentioned by @rgommers:

```python
from simsimd import pointer_to_sqeuclidean, pointer_to_cosine, pointer_to_inner
```

I chose vanilla integers (to cast pointers) as opposed to Python capsules for usability, but I can add alternative interfaces as well.

**Comment by tylerjereddy** [comment]
Seems like an interesting proposal. I agree with above comments about trying to coordinate with sklearn where reasonable, and of course there's the usual concern about extra complexity/maintenance burden.

I make pretty heavy use of `pdist`/`cdist` and I suspect many others do too, so keeping those the same (or better) is important I think.

I do wonder if there's any concern about the sample call signature above `d = cdist(x, y, 'euclidean') ` being indistinguishable from the "native SciPy" call signature with respect to reasoning/clarity. For example, is there merit to having the "explicit is better than implicit" philosophy on the name of the metric at least telling us that it has an external provider like `euclidean_tpl` for the third party lib or something like that.

**Comment by fancidev** [comment]
> I chose vanilla integers (to cast pointers) as opposed to Python capsules for usability, but I can add alternative interfaces as well.

Thanks for the heads up! I’m working on a prototype to try this out.

> For example, is there merit to having the "explicit is better than implicit" philosophy on the name of the metric at least telling us that it has an external provider like `euclidean_tpl` for the third party lib or something like that.

Good point! I’m going to prototype a “new_cdist” that takes an optional backend argument that allows the user to explicitly specify a backend. The backend may also use a metric name like `euclidean_tpl` if it choose to.

**Comment by Micky774** [comment]
Thank you for taking the time to create this discussion! I'm definitely +1 for a unified plug-in mechanism. It definitely is the better "upstream" method to implement a lot of the work I've tried nudging towards in scikit-learn while compartmentalizing the complexity.

I'm wary of how we handle the extra args, especially for cases where you have multiple extra args for a common function implemented via multiple backends, yet the order of the args are permuted in the different backends. If an implementation doesn't thoroughly validate its extra args (which IMO should be a python-level task not a C-level one) then this could cause issues that are potentially opaque to end users, especially since one can simply say it was user error for assuming an incorrect API. Most notable, it would prevent a real "drop-in" solution.

This would mainly be an issue, I imagine, in areas where SciPy doesn't have an established API (e.g. new metrics / kernels) and multiple libraries implement differing APIs for the same semantic function. This isn't a huge area right now, but w/ the adoption of this framework, it may be more likely to be an issue later. It may just have to be a "buyer beware" situation where folks should be aware to individually validate the APIs of the backends they use, in case they differ from what they're expecting.

With that being said, the presence of the `void*` args would allow for "functional" or "fused" kernels which complete separate computation in a way that is local to the specific distance calculations, optimizing for cache efficiency, memory movement, and vtable lookups -- scikit-learn leverages this quite heavily in the most performance critical sections thanks to @jjerphan's system. This wouldn't be something end users would be expected to care about -- or see in the first place -- but it would be a nice tool for library developers.

TL;DR
- I think this would be valuable to the ecosystem.
- I think establishing the API in SciPy makes sense due to its ubiquity. 
- I would be interested in hearing whether other folks are concerned about inter-backend API compatibility regarding `void*` args
- I would be interested in seeing whether existing scikit-learn fused kernels can be re-contextualized in terms of the proposed API

**Comment by lucascolley** [comment]
> It may just have to be a "buyer beware" situation where folks should be aware to individually validate the APIs of the backends they use, in case they differ from what they're expecting.

Yeah, this is one of those aspects that is on the level of generality of dispatching ideas like `spatch` I linked to above. I guess it would be nice to try to be consistent with any efforts like that.

**Comment by ashvardanian** [comment]
I think it's common for power users to encounter "buyer beware" situations, as seen with SciPy and NumPy, where performance varies based on the detected BLAS implementation.

---

This proposal opens an opportunity to refactor SciPy’s distances module. There are several issues I've noticed in the last months, and I'm curious if the community agrees.

1. **Weighted Functions Usage**:  
   Weighted variants are conceptually interesting, but rarely used. This also complicates the API since some distances, like Jensen-Shannon and Mahalanobis, don’t support weighting.
   
2. **Irrelevant Kernels**:  
   Many documented distance functions, such as Kulczynski1 and Sokal-Michener, are unpopular. There are overlaps, like `cosine` calling `correlation`. Also, `euclidean` is a trivial extension of `sqeuclidean` distance. Supporting all of those variants across all SIMD generations is troublesome. Smaller palette of more optimized kernels may be a good idea.

3. **Mixed Precision**:  
   SciPy kernels operating on small data types (`i8`, `i16`, etc.) may overflow. Upcasting inputs can avoid errors without breaking the API. Implementing mixed-precision optimizations could improve robustness.

4. **Signature and Input Representation**:  
   Python lacks descriptors for types like `bf16` or compressed bit-matrix formats (`np.packbits`), which are useful for boolean distances. SciPy's reliance on integers for booleans limits its performance. A `dtype=` option for specialized data would help address this.

**Comment by tylerjereddy** [comment]
> Many documented distance functions, such as Kulczynski1 and Sokal-Michener, are unpopular. 

Some potential good news there, both were deprecated in gh-21572, and we coordinated with sklearn such that they may also follow suit.

**Comment by fancidev** [comment]
Thanks for the comments @Micky774 !

> I'm wary of how we handle the extra args, especially for cases where you have multiple extra args for a common function implemented via multiple backends, yet the order of the args are permuted in the different backends.

What I have in mind is to (1) let the backend declare the order, name, and dimension of additional parameters it supports, (2) let the user call cdist passing additional arguments as keyword-only, and (3) let the front end validate the user supplied arguments.

To prevent “combinatorial explosion”, I do require additional parameters to (1) have the same dtype as `out` (converting user input if necessary), and (2) have shape `()`, `(n,)`, or `(n,n)` where n is the vector dimension.

> I would be interested in seeing whether existing scikit-learn fused kernels can be re-contextualized in terms of the proposed API.

I’m not too familiar with scikit-learn. Could you elaborate a bit what is “fused kernels”?

**Comment by fancidev** [comment]
Those are interesting points! @ashvardanian 

> Weighted variants are conceptually interesting, but rarely used. 

I too guess they’re not widely used. At least I’ve not see a weight parameter in the distance function of Scikit-Learn, MATLAB, Mathematica, STATA.

> This also complicates the API since some distances, like Jensen-Shannon and Mahalanobis, don’t support weighting.

The backend doesn’t have to support weighting if it doesn’t bother to. Though technically, one could define weights for Jensen-Shannon reasonably, and the inverse covariance matrix of Mahalanobis could be considered a weight matrix itself (in other words, Mahalanobis could be interpreted as matrix-weighted Euclidean).

> Also, `euclidean` is a trivial extension of `sqeuclidean` distance.

I think `euclidean` is more subtle than `sqeuclidean`, because the obvious calculation (taking square root of sum of squares) may give an overflow result when the correct result does not overflow, and a more careful calculation that avoids this problem is likely slower. This is the kind of trade-off that a backend has got to offer for the “power user”.

> Supporting all of those variants across all SIMD generations is troublesome. Smaller palette of more optimized kernels may be a good idea.

I agree!

> SciPy kernels operating on small data types (`i8`, `i16`, etc.) may overflow.

I believe SciPy upcasts all integral input to float64 before calculation. There is room for performance improvement if the input is integral (e.g. SimSIMD can compute cosine distance from i8 input). Some details need to be worked out for the (to-be) front-end with respect to input type conversion.

> Python lacks descriptors for types like `bf16` or compressed bit-matrix formats (`np.packbits`), which are useful for boolean distances

Some work can be done by the front-end to “recognize” bf16 or bitmask input. For example, one might call `pdist(BFloat16(x))` to let the front-end “interpret” a 16-bit array as containing bf16 numbers. Similar enhancements could come as follow ups.

**Comment by fancidev** [comment]
I added a prototype (gh-21669) to illustrate how the proposed interfaces may work in practice.

TL DR:

The minimal working code (from the Python side) to register and use a kernel is the following:

```Python
import scipy
import scipy.spatial.distance as sd
capsule = sd._distance_wrap.cosine_DistanceMatrix_capsule()
kernel = scipy.LowLevelCallable(capsule)
sd.add_distance_kernel('cosine.m', 'f64', kernel, 'alt')
sd.cdist([[1, 2, 3]], [[4, 5, 6]], 'new_cosine', backend='alt')
```

It could be helpful to check the [rendered docstring](https://output.circle-artifacts.com/output/job/d037d9f3-7cab-405d-9171-d3b889d81955/artifacts/0/html/reference/generated/scipy.spatial.distance.add_distance_kernel.html) of `add_distance_kernel ` to get more details on how to register a distance computation kernel.

**Comment by Micky774** [comment]
> I’m not too familiar with scikit-learn. Could you elaborate a bit what is “fused kernels”?

Here I refer to algorithms where we compute pairwise distances on feature data **and** produce a side-effect, such as `ArgKminClassMode`, a scikit-learn computation backend class which computes the pairwise distances **and** keeps a running score of the class label distribution for each points k-neighbors for use in `KNeighborsClassifier` as its side-effect. Performing both simultaneously (or close enough) is that we can guarantee the computed distances are only introduced into the cache once, so we spare ourselves quite a lot of memory transfer costs.

**Comment by jjerphan** [comment]
To complete Meekail's comment, I would recommend reading [those high-level pieces of documentations of `PairwiseDistancesReductions`](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/metrics/_pairwise_distances_reduction/__init__.py) and subsequent elements of documentations which are closer to the implementation in the same directory.
